cherab/tools/observers/bolometry.py

```diff
@@ -687,6 +695,350 @@ def etendue_single_run(_):
+
+        sightlines = []
+        for pixel_column in self.pixels_as_foils:
+            sightline_column = []
+            for pixel in pixel_column:
+                los_observer = SightLine(pipelines=[pipeline(accumulate=False)],
+                                         pixel_samples=1, quiet=True,
+                                         parent=self, name=pixel.name + "Sightline")
+                los_observer.render_engine = self.render_engine
+                los_observer.spectral_bins = self.spectral_bins
+                los_observer.min_wavelength = self.min_wavelength
+                los_observer.max_wavelength = self.max_wavelength
+                sightline_column.append(los_observer)
+            sightlines.append(sightline_column)
+
+        return np.asarray(sightlines, dtype='object')
+
+    def trace_sightlines(self, rows="all", columns="all"):
+        """
+        Traces the central sightlines through each pixel in the detector to see where the sightline terminates.
+
+        Raises a RuntimeError exception if no intersections were found.
+
+        :param rows: which pixel rows to trace. Can be an integer, a
+          list of integers or "all".
+        :param columns: which pixel columns to trace. Can be an
+          integer, a list of integers or "all".
+
+        :return: A 2D array of tuples containing the origin point, hit
+          point and terminating surface primitive for each pixel.
+        """
+        if rows == "all":
+            rows = slice(None)
+        if columns == "all":
+            columns = slice(None)
+        pixels = np.atleast_2d(self.pixels_as_foils[columns, rows])
+        traces = [[pixel.trace_sightline() for pixel in pixel_column] for pixel_column in pixels]
+        return np.asarray(traces, dtype='object')
+
+    def calculate_sensitivity(self, voxel_collection, ray_count=10000):
+        r"""
+        Calculates a sensitivity vector for this detector on the specified voxel collection.
+
+        This function is used for calculating sensitivity matrices which can be combined for
+        multiple detectors into a sensitivity matrix :math:`\mathbf{W}`.
+
+        :param VoxelCollection voxel_collection: The voxel collection on which to calculate
+          the sensitivities.
+        :param int ray_count: The number of rays to use in the calculation. This should be
+          at least >= 10000 for decent statistics.
+        :return: A 3D array of sensitivities (ncol, nrow, nvoxels)
+        """
+        # This method exploits ToroidalVoxelCollection.set_active("all"), which
+        # makes each voxel emit a different wavelength of light. By observing
+        # the voxel collection with a spectral pipeline we can thus distinguish
+        # the amount of emission from each individual voxel.
+        if not isinstance(voxel_collection, VoxelCollection):
+            raise TypeError("voxel_collection must be of type VoxelCollection")
+
+        if self.units == "Power":
```

Comment:
3rd appearance of an messy bit of string handling. Apply DRY (Don't Repeat Yourself).

Reply:
Done